src/transport_udp.cpp

```diff
@@ -242,361 +520,480 @@ UDPTransport::fd_writer()
+        logger->Log(cantina::LogLevel::Critical, s_log.str());
+        throw std::runtime_error(s_log.str());
+    }
 
-  connData cd;
-  cd.data = bytes;
-  cd.contextId = context_id;
-  cd.streamId = streamId;
 
-  if (not fd_write_queue.push(cd)) {
-    return TransportError::QueueFull;
-  }
+    struct sockaddr_in srvAddr;
+    srvAddr.sin_family = AF_INET;
+    srvAddr.sin_addr.s_addr = htonl(INADDR_ANY);
+    srvAddr.sin_port = 0;
+    err = bind(fd, (struct sockaddr *) &srvAddr, sizeof(srvAddr));
+    if (err) {
+        s_log << "client_connect: Unable to bind to socket: " << strerror(errno);
+        logger->Log(cantina::LogLevel::Critical, s_log.str());
+        throw std::runtime_error(s_log.str());
+    }
 
-  return TransportError::None;
-}
+    std::string sPort = std::to_string(htons(serverInfo.port));
+    struct addrinfo hints = {}, *address_list = NULL;
+    hints.ai_family = AF_INET;
+    hints.ai_socktype = SOCK_DGRAM;
+    hints.ai_protocol = IPPROTO_UDP;
+    err = getaddrinfo(
+            serverInfo.host_or_ip.c_str(), sPort.c_str(), &hints, &address_list);
+    if (err) {
+        strerror(1);
+        s_log << "client_connect: Unable to resolve remote ip address: "
+              << strerror(errno);
+        logger->Log(cantina::LogLevel::Critical, s_log.str());
+        throw std::runtime_error(s_log.str());
+    }
 
-std::optional<std::vector<uint8_t>>
-UDPTransport::dequeue(const TransportConnId& context_id,
-                      const DataContextId& streamId)
-{
+    struct addrinfo *item = nullptr, *found_addr = nullptr;
+    for (item = address_list; item != nullptr; item = item->ai_next) {
+        if (item->ai_family == AF_INET && item->ai_socktype == SOCK_DGRAM &&
+            item->ai_protocol == IPPROTO_UDP) {
+            found_addr = item;
+            break;
+        }
+    }
 
-  if (remote_contexts.count(context_id) == 0) {
-    logger->warning << "dequeue: invalid context id: " << context_id
-                    << std::flush;
-    // Invalid context id
-    return std::nullopt;
-  }
+    if (found_addr == nullptr) {
+        logger->critical << "client_connect: No IP address found" << std::flush;
+        throw std::runtime_error("client_connect: No IP address found");
+    }
 
-  if (dequeue_data_map[context_id].count(streamId) == 0) {
-    logger->error << "dequeue: invalid stream id: " << streamId << std::flush;
+    struct sockaddr_in *ipv4 = (struct sockaddr_in *) &serverAddr.addr;
+    memcpy(ipv4, found_addr->ai_addr, found_addr->ai_addrlen);
+    ipv4->sin_port = htons(serverInfo.port);
+    serverAddr.addr_len = sizeof(sockaddr_in);
 
-    return std::nullopt;
-  }
+    freeaddrinfo(address_list);
 
-  auto& dq = dequeue_data_map[context_id][streamId];
+    serverAddr.id = create_addr_id(serverAddr.addr);
 
-  if (dq.size() <= 0) {
-    return std::nullopt;
-  }
+    ++last_conn_id;
 
-  return dq.pop().value().data;
-}
+    const auto& [conn_it, _] = conn_contexts.emplace(last_conn_id,
```

Comment:
Is a lock needed to protect this map?

Reply:
Not on this one because it's creating the client.  Although, we can add a lock for the crazy use-case where the app calls start and close at the same time from two different threads.

Reply:
While not needed, consolidating the send/recv lock to connections lock will result in this being protected by lock as well.